File: example-simulator.php
<?php

error_reporting(E_ALL);

require_once 'BattleEngine.php';
require_once 'OG.php';

use BattleEngine\BattleEngine;
use BattleEngine\Combatant;
use OG\OG;

$numSimulations = 100;

$attackers = [
    new Combatant(
        10,
        10,
        10,
        [
            OG::BATTLESHIP => 25000,
        ],
    ),
    new Combatant(
        10,
        10,
        10,
        [
            OG::LIGHT_FIGHTER => 25000,
            OG::HEAVY_FIGHTER => 25000,
            OG::CRUISER => 25000,
        ],
    ),
];

$defenders = [
    new Combatant(
        10,
        10,
        10,
        [
            OG::DEATH_STAR => 250,
        ],
    ),
];

function printCombatant(Combatant $combatant)
{
    $weapons = $combatant->getWeaponsTechnology() * 10;
    $shield = $combatant->getShieldingTechnology() * 10;
    $armor = $combatant->getArmorTechnology() * 10;
    printf(
        "\tWeapons %d%% Shield %d%% Armor %d%%\n",
        $weapons,
        $shield,
        $armor
    );
    $unitGroups = $combatant->getUnitGroups();
    foreach ($unitGroups as $kind => $count) {
        printf("\t%s %d\n", OG::$names[$kind], $count);
    }
}

function printCombatants(string $who, array $combatants)
{
    foreach ($combatants as $i => $combatant) {
        printf("%s #%d\n", $who, $i);
        printCombatant($combatant);
        printf("\n");
    }
}

function mean(array $values): float
{
    return array_sum($values) / count($values);
}

function stdev(array $values): float
{
    $mean = array_sum($values) / count($values);
    $var = 0.0;
    foreach ($values as $val) {
        $var += pow($val - $mean, 2);
    }
    return sqrt($var / (count($values) - 1));
}

function printCombatantResult(Combatant $combatant, array $stats)
{
    $unitGroups = $combatant->getUnitGroups();
    foreach ($unitGroups as $kind => $count) {
        printf("\t%s:\n", OG::$names[$kind]);
        $numRemainingUnits = array_map(fn($s) => $s[$kind]->getNumRemainingUnits(), $stats);
        printf("\t\tMean:  %f\n", mean($numRemainingUnits));
        printf("\t\tStdev: %f\n", stdev($numRemainingUnits));
        printf("\t\tMin:   %d\n", min($numRemainingUnits));
        printf("\t\tMax:   %d\n", max($numRemainingUnits));
    }
}

function printCombatantResults(string $who, array $combatants, array $simulations, callable $getOutcomes)
{
    foreach ($combatants as $i => $combatant) {
        printf("%s #%d\n", $who, $i);
        $stats = array_map(fn($s) => $getOutcomes($s)[$i]->getRoundStats($s->getNumRounds() - 1), $simulations);
        printCombatantResult($combatant, $stats);
        printf("\n");
    }
}

printCombatants('Attacker', $attackers);
printCombatants('Defender', $defenders);

$battleEngine = new BattleEngine('./build/BattleEngine', OG::$unitsAttributes);
$simulations = $battleEngine->simulate($attackers, $defenders, 0, $numSimulations);

$numRounds = mean(array_map(fn($s): int => $s->getNumRounds(), $simulations));
printf("Num rounds: %f\n\n", $numRounds);
printCombatantResults('Attacker', $attackers, $simulations, fn($s) => $s->getAttackersOutcomes());
printCombatantResults('Defender', $defenders, $simulations, fn($s) => $s->getDefendersOutcomes());

File: example-simulator.py
import statistics
from typing import Dict, List

import OG
from BattleEngine import BattleEngine, BattleOutcome, Combatant, UnitGroupStats, UnitKind

num_simulations = 100

attackers = [
    Combatant(
        weapons_technology=10,
        shielding_technology=10,
        armor_technology=10,
        unit_groups={
            OG.Battleship: 25000,
        },
    ),
    Combatant(
        weapons_technology=10,
        shielding_technology=10,
        armor_technology=10,
        unit_groups={
            OG.LightFighter: 25000,
            OG.HeavyFighter: 25000,
            OG.Cruiser: 25000,
        },
    ),
]

defenders = [
    Combatant(
        weapons_technology=10,
        shielding_technology=10,
        armor_technology=10,
        unit_groups={
            OG.DeathStar: 250,
        },
    ),
]


def print_combatant(combatant: Combatant):
    weapons = combatant.weapons_technology * 10
    shield = combatant.shielding_technology * 10
    armor = combatant.armor_technology * 10
    print('\tWeapons {}% Shield {}% Armor {}%'.format(weapons, shield, armor))
    for kind, count in combatant.unit_groups.items():
        print('\t{}\t{}'.format(OG.names[kind], count))


def print_combatants(who: str, combatants: List[Combatant]):
    for i, combatant in enumerate(combatants):
        print('{} #{}'.format(who, i))
        print_combatant(combatant)
        print()


def print_combatant_result(combatant: Combatant, stats: List[Dict[UnitKind, UnitGroupStats]]):
    for kind in combatant.unit_groups.keys():
        print('\t{}:'.format(OG.names[kind]))
        num_remaining_units = [s[kind].num_remaining_units for s in stats]
        print('\t\tMean:  {}'.format(statistics.mean(num_remaining_units)))
        print('\t\tStdev: {}'.format(statistics.pstdev(num_remaining_units)))
        print('\t\tMin:   {}'.format(min(num_remaining_units)))
        print('\t\tMax:   {}'.format(max(num_remaining_units)))


def print_combatant_results(who: str, combatants: List[Combatant], simulations: List[BattleOutcome],
                            get_outcomes):
    for i, combatant in enumerate(combatants):
        print('{} #{}'.format(who, i))
        stats = [get_outcomes(s)[i].round_stats(s.num_rounds - 1)
                 for s in simulations]
        print_combatant_result(combatant, stats)
        print()


print_combatants('Attacker', attackers)
print_combatants('Defender', defenders)

engine = BattleEngine('./build/BattleEngine', OG.units_attributes)
simulations = engine.simulate(attackers, defenders, 0, num_simulations)

num_rounds = statistics.mean(s.num_rounds for s in simulations)
print('Num rounds: {}\n'.format(num_rounds))
print_combatant_results('Attacker', attackers, simulations,
                        lambda s: s.attackers_outcomes)
print_combatant_results('Defender', defenders, simulations,
                        lambda s: s.defenders_outcomes)
